Add bfloat16 support to JAX.

hawkinsp opened the pull request:
bfloat16 support is still immature, but this PR adds some initial support.

Fixes #76, at least enough that we can declare it fixed and open specific issues for specific bfloat16 problems.

The main awkwardness that this change deals with is that classic NumPy doesn't understand bfloat16 promotion rules, so we must:
* implement our own type promotion operators that understand bfloat16 types
* wrap a number of the reference implementations in tests to temporarily cast to float32 for computation.

hawkinsp committed:
Add some initial bfloat16 support to JAX.

mattjj commented on jax/numpy/lax_numpy.py:
```diff
@@ -1123,16 +1131,18 @@ def reduction(a, axis=None, dtype=None, out=None, keepdims=False):
-    if _dtype(a) != result_dtype:
-      a = lax.convert_element_type(a, result_dtype)
+    dtype = dtype or _dtype(a)
+    #result_dtype = _dtype(np_fun(onp.ones((), dtype=dtype or _dtype(a))))
```
Don't need this line anymore?

mattjj commented on jax/dtypes.py:
```diff
@@ -96,13 +116,58 @@ def coerce_to_array(x):
 can_cast = onp.can_cast
-issubdtype = onp.issubdtype
 issubsctype = onp.issubsctype
 promote_types = onp.promote_types
```
Remove this, since it's now shadowed by the definition below?

mattjj reviewed (approved):
Awesome. Closing an issue from Dec 2018!

Not necessarily as part of this PR, but since your head is in this space now, I wonder: is it clear to you where we can add memoization to speed up dtype promotion logic?

hawkinsp committed:
Fix test failures.

To fix some accuracy problems with some reduction tests, rework handling of types in several reductions.

Allow lax.add and lax.mul to support bool types to ease reduction implementation.

hawkinsp committed:
Revert addition of add/mul translation rules for bools.

hawkinsp committed:
Relax test tolerances further.

hawkinsp committed:
Relax test tolerances.

hawkinsp committed:
Take both arguments and result types into account for tolerance.

hawkinsp committed:
Add comment to dtypes.py.

hawkinsp committed:
Fix TPU default tolerance.
